### DISCUSSION TITLE
How to open the application automatically after boot? (Android 10)

### DISCUSSION THREAD

**Discussion by oguzhan51511** [dotnet/maui#10414] [Q&A]
I am able to display toast message in BootReceiver. I cannot start the application. I managed to autostart on a device with android api 25, but the target device is Android api 29. Any ideas on this?

How can I start the application automatically when Android boot complate?

**BootReceiver.cs**
-----------------
using Android.App;
using Android.Content;
using Android.Widget;

namespace MaRadio.Platforms.Android
{
    [BroadcastReceiver(Label = "BootReceiver", DirectBootAware = true, Enabled = true, Exported = true)]
    [IntentFilter(new[] { Intent.ActionBootCompleted }, Priority = (int)IntentFilterPriority.HighPriority)]

    public class BootReceiver : BroadcastReceiver
    {
        public override void OnReceive(Context context, Intent intent)
        {
            Toast.MakeText(context, " Service Running ", ToastLength.Long)?.Show();
            var launch_intent = Platform.CurrentActivity?.PackageManager?.GetLaunchIntentForPackage(Platform.CurrentActivity.PackageName);
            if (launch_intent != null)
            {
                launch_intent.AddFlags(ActivityFlags.ReorderToFront);
                launch_intent.AddFlags(ActivityFlags.NewTask);
                launch_intent.AddFlags(ActivityFlags.ResetTaskIfNeeded);
                Platform.CurrentActivity?.StartActivity(launch_intent);
            }
        }
    }
}

**AndroidManifest.xml**
-----------------
android.permission.RECEIVE_BOOT_COMPLETED
android.permission.SYSTEM_ALERT_WINDOW
android.permission.ACCESS_NETWORK_STATE
android.permission.INTERNET

**Comment by mgungorchamp** [comment]
Did you find the solution?

**Reply by oguzhan51511** [reply]
**Platforms/Android/AndroidManifest.xml**
```
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android">
	<application android:allowBackup="true"
               android:icon="@mipmap/appicon"
               android:roundIcon="@mipmap/appicon_round"
               android:supportsRtl="false"></application>
  <uses-permission android:name="android.permission.SYSTEM_ALERT_WINDOW"/>
  <uses-permission android:name="android.permission.INTERNET"/>
  <uses-permission android:name="android.permission.REQUEST_INSTALL_PACKAGES"/>
  <uses-permission android:name="android.permission.INSTALL_PACKAGES" />
  <uses-permission android:name="android.permission.DELETE_PACKAGES" />
  <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED"/>
  <uses-permission android:name="android.permission.ACCESS_WIFI_STATE"/>
  <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE"/>
  <uses-permission android:name="android.permission.WAKE_LOCK"/>
  <uses-permission android:name="android.permission.FOREGROUND_SERVICE"/>
</manifest>
```

**Platforms/Android/BootReceiver.cs**

```
using Android.App;
using Android;
using Android.Content;
using Android.Widget;

[assembly: UsesPermission(Manifest.Permission.ReceiveBootCompleted)]

namespace MaRadio.Platforms.Android
{
    [BroadcastReceiver(Name = "com.oskay.maradio.BootReceiver", Exported = true, Enabled = true, DirectBootAware = true)]
    [IntentFilter(new[] {  Intent.ActionReboot,
        Intent.ActionBootCompleted
        , Intent.ActionLockedBootCompleted
        , Intent.ActionMyPackageReplaced
        , Intent.ActionUserInitialize
        , "android.intent.action.QUICKBOOT_POWERON"
        , "com.htc.intent.action.QUICKBOOT_POWERON" }, Categories = new[] { Intent.CategoryDefault })]
    public class BootReceiver : BroadcastReceiver
    {
        public override void OnReceive(Context context, Intent intent)
        {
            if (intent.Action == null || (!intent.Action.Equals(Intent.ActionBootCompleted)
                                          && !intent.Action.Equals(Intent.ActionReboot)
                                          && !intent.Action.Equals(Intent.ActionLockedBootCompleted)
                                          && !intent.Action.Equals(Intent.ActionUserInitialize)
                                          && !intent.Action.Equals("android.intent.action.QUICKBOOT_POWERON")
                                          && !intent.Action.Equals("com.htc.intent.action.QUICKBOOT_POWERON"))) return;
            Toast.MakeText(context, "maRadio Starting...", ToastLength.Long)?.Show();
            var s = new Intent(context, typeof(MainActivity));
            s.AddFlags(ActivityFlags.ResetTaskIfNeeded);
            s.AddFlags(ActivityFlags.ReorderToFront);
            s.AddFlags(ActivityFlags.NewTask);
            context.StartActivity(s);
        }
    }
}
```

**Platforms/Android/MainActivity.cs**

```
using Android.App;
using Android.Content.PM;
using Android.OS;

namespace MaRadio;

[Activity(Theme = "@style/Maui.SplashTheme",
    Enabled = true,
    Exported = true,
    MainLauncher = true,
    LaunchMode = LaunchMode.SingleTask,
    ConfigurationChanges =  ConfigChanges.ScreenSize | ConfigChanges.Orientation | ConfigChanges.UiMode | ConfigChanges.ScreenLayout | ConfigChanges.SmallestScreenSize | ConfigChanges.Density)]
[IntentFilter(new[] { Platform.Intent.ActionAppAction }, Categories = new[] { global::Android.Content.Intent.CategoryLauncher })]
public class MainActivity : MauiAppCompatActivity
{
}
```

**Reply by mgungorchamp** [reply]
Thank you very much. Worked perfectly! 
Cok tesekkurler

**Reply by mgungorchamp** [reply]
Another Question: I am not sure it's related to this: My app perfectly starts at boot, thank you again, using Amazon Fire TV HDMI Stick, but after a while, the screen goes blank, like energy saver mode, how can I make it on all the time programmatically, since Fire TV does not have any option to prevent that.

**Reply by mgungorchamp** [reply]
I think I found it:  
`<uses-permission android:name="android.permission.FLAG_KEEP_SCREEN_ON"/>`

**Reply by oguzhan51511** [reply]
Using the KeepScreenOn Property: In MAUI, you can prevent the screen from turning off by using the KeepScreenOn property on a view derived from the Page or ContentView class. For example, you can set the KeepScreenOn property for a ContentPage as follows:

``using Microsoft.Maui.Controls;

public class MyContentPage : ContentPage
{
    public MyContentPage()
    {
        // Set the KeepScreenOn property to true to prevent the screen from turning off
        KeepScreenOn = true;
    }
}
``

By doing this, when MyContentPage is used, the screen will stay on.

Platform-Specific Customizations: If you need more control and customization, you can make platform-specific customizations. For example, in Android, you can make customizations in the MainActivity.cs file, as shown in the following example:

``using Microsoft.Maui;
using Microsoft.Maui.Controls.PlatformConfiguration.AndroidSpecific;
using Microsoft.Maui.Controls.PlatformConfiguration;
using Application = Microsoft.Maui.Controls.Application;

namespace MyMauiApp
{
    public class MainActivity : MauiAppCompatActivity
    {
        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);

            // Platform-specific customization to prevent the screen from turning off
            this.On<Android>().UseWindowSoftInputModeAdjust(WindowSoftInputModeAdjust.Resize);
            this.On<Android>().KeepScreenOn();

            // Other necessary code...
        }
    }
}
``

In this example, we're using the KeepScreenOn() method in the OnCreate method of the MainActivity class to prevent the screen from turning off.

These methods are two common approaches to preventing the screen from turning off in a MAUI application using C# and .NET MAUI. Depending on your requirements, you can choose the method that best suits your application

**Reply by mgungorchamp** [reply]
Thank you again,
I needed to use `DeviceDisplay.Current.KeepScreenOn = true;   `  otherwise it did not recognize KeepScreenOn parameter.  I hope I am doing the right thing.

**Reply by mgungorchamp** [reply]
Another Question: When I update the app or reinstall, I do not want it to wipe out /delete the files under the  `FileSystem.Current.AppDataDirectory`  is it related to these settings?

**Comment by HuaFangYun** [comment]
Can you provide a demo program? Thank you!

**Reply by mgungorchamp** [reply]
https://github.com/mgungorchamp/InfoBoard

**Reply by HuaFangYun** [reply]
thanks
